Add feature for the stalker exclusion when BLR

akiannillo opened the pull request:
I modified the virtualize branch instruction function in the stalker, added two tests, and a script to run test with android arm 64.

I have only a doubt whether to add something for the unfollow me, such in [L554](https://github.com/frida/frida-gum/blob/master/gum/backend-arm64/gumstalker-arm64.c#L554).

(new performance duration ratio is at most 1.15)

oleavr commented on gum/backend-arm64/gumstalker-arm64.c:
```diff
@@ -2673,6 +2691,36 @@ gum_exec_block_write_call_invoke_code (GumExecBlock * block,
+  }
+}
+
+static GumAddress
```
- The `is_` part of this name makes it sound like it's returning a `gboolean` – so I suppose that would be clearer?
- This is pretty high-level, so perhaps it would be good to move this right after `gum_stalker_exclude`? (Which also means we don't have to put a forward-declaration.)

oleavr commented on gum/backend-arm64/gumstalker-arm64.c:
```diff
@@ -2479,8 +2483,13 @@ gum_exec_block_write_call_invoke_code (GumExecBlock * block,
+  GumBranchTarget next_insn_as_target = { 0, };
+  next_insn_as_target.absolute_address = gc->instruction->end;
+  next_insn_as_target.reg = ARM64_REG_INVALID;
+
```
Would be good to move this down to the local scope (as it's only used there).

oleavr commented on gum/backend-arm64/gumstalker-arm64.c:
```diff
@@ -2570,6 +2579,15 @@ gum_exec_block_write_call_invoke_code (GumExecBlock * block,
+        GUM_ARG_REGISTER, ARM64_REG_X15);
+
+    gum_arm64_writer_put_mov_reg_reg (cw, ARM64_REG_X15, ARM64_REG_X0);
+    gum_arm64_writer_put_cbz_reg_label (cw, ARM64_REG_X15, keep_this_blr);
```
So if we change it to return a `gboolean` that will simplify this logic a bit.

akiannillo replied:
That was a trick (actually just to save an instruction). In the case of valid register not excluded, we need the target address in register X15. We should have done this by pushing and popping X15 around the `gum_stalker_is_address_included`call. With this solution we directly use the return value of the function by moving X0 into X15, that is 0 (false) when the address need to be excluded.
I know it's micro-optimization, but if you think it's clearer use a gboolean return value, I'll change it.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,142 @@ store_x0 (GumCpuContext * cpu_context,
+static void
+hello_ken (void)
+{
+  g_print ("hello ken! ");
```
Tests shouldn't print anything unless it's really needed. Perhaps just a `g_usleep (1)` instead?

akiannillo replied:
I was initially thinking about `g_usleep (1)`, but I wanted an immediate feedback that the function is actually invoked. 
Now I see no reason why the function should not be called. I'm changing it.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,142 @@ store_x0 (GumCpuContext * cpu_context,
+}
+
+static void
+count_instructions (GumStalkerIterator * iterator,
```
For consistency with the other tests it would be good to instead enable the `EXEC` events and look at the length of the events on the fake eventsink.

oleavr commented on tests/run-android-64.sh:
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
```
Perhaps naming this `run-android-arm64` would be better? We can rename the other one also while at it.

oleavr commented on gum/backend-arm64/gumstalker-arm64.c:
```diff
@@ -1016,6 +1019,7 @@ static guint total_transitions = 0;
 
 GUM_DEFINE_ENTRYGATE (call_imm)
 GUM_DEFINE_ENTRYGATE (call_reg)
+GUM_DEFINE_ENTRYGATE (jmp_over_call_reg)
```
Maybe `call_reg_excluded` would be clearer?

oleavr commented:
Oh, and did you test calling `unfollow()` from inside an excluded call? Probably worth having test-coverage for this.

akiannillo commented:
I added the tests to test calling `unfollow()` from inside an excluded call, and I changed the function that check the inclusion/exclusion of the address to modify also the boolean value has_call_to_excluded_range. It actually became a `gum_exec_block_' function, and I changed the name so that it does not sound as a predicate.

oleavr commented on gum/backend-arm64/gumstalker-arm64.c:
```diff
@@ -322,6 +322,8 @@ static GumExecBlock * gum_exec_block_new (GumExecCtx * ctx);
 static GumExecBlock * gum_exec_block_obtain (GumExecCtx * ctx,
     gpointer real_address, gpointer * code_address);
 static gboolean gum_exec_block_is_full (GumExecBlock * block);
+static GumAddress gum_exec_block_check_address_for_exclusion (GumExecBlock * block,
```
Exceeds 80 cols.

oleavr commented on gum/backend-arm64/gumstalker-arm64.c:
```diff
@@ -2024,6 +2027,26 @@ gum_exec_block_is_full (GumExecBlock * block)
+    if (GUM_MEMORY_RANGE_INCLUDES (r, address))
+    {
+      block->has_call_to_excluded_range = TRUE;
+      return GUM_ADDRESS(0);
```
Missing space before `(`.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+
+  func = GUM_POINTER_TO_FUNCPTR (StalkerTestFunc,
+      test_arm64_stalker_fixture_dup_code (fixture, code_template,
+        sizeof (code_template)));
```
Each level of hanging indent should be 4 spaces. The outermost here is 2.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+    0xd2801320, /* mov x0, #0x99  */
+    0xd65f03c0, /* ret            */
+  };
+
```
This blank line could be dropped.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+
+  func_a_address = fixture->code + (16 * 4);
+
+  GumMemoryRange memory_range;
```
Declaration after statement.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+  guint8 * func_a_address;
+
+  fixture->sink->mask = GUM_EXEC;
+  g_assert_cmpuint (fixture->sink->events->len, ==, 0);
```
Let's move these two lines down to right before `follow_and_invoke()`, to make the logic at the end clearer.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+  StalkerTestFunc func;
+  guint8 * code;
+  GumArm64Writer cw;
+  gint r;
```
Let's avoid single-letter variable names. Maybe name it `result` or something descriptive.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+  gum_clear_cache (cw.base, gum_arm64_writer_offset (&cw));
+  gum_arm64_writer_clear (&cw);
+
+  GumMemoryRange memory_range;
```
Declaration after statement.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+
+  GumMemoryRange memory_range;
+  memory_range.base_address = GUM_ADDRESS (func_a);
+  memory_range.size = (4 * 2);
```
Redundant parentheses (and they don't add clarity).

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+
+  GumMemoryRange memory_range;
+  memory_range.base_address = (GumAddress) func_a_address;
+  memory_range.size = (4 * 2);
```
Redundant parentheses (and they don't add clarity).

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+  const gchar * start_lbl = "start";
+
+  fixture->sink->mask = GUM_EXEC;
+  g_assert_cmpuint (fixture->sink->events->len, ==, 0);
```
Let's move these two lines down to right before `func ()`, to make the logic at the end clearer.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+  gum_clear_cache (cw.base, gum_arm64_writer_offset (&cw));
+  gum_arm64_writer_clear (&cw);
+
+  GumMemoryRange memory_range;
```
Declaration after statement.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+
+  GumMemoryRange memory_range;
+  memory_range.base_address = GUM_ADDRESS (func_a);
+  memory_range.size = (4 * 20);
```
Redundant parentheses (and they don't add clarity).

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+  StalkerTestFunc func;
+  guint8 * code;
+  GumArm64Writer cw;
+  gint r;
```
Let's avoid single-letter variable names. Maybe name it `result` or something descriptive.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+  StalkerTestFunc func;
+  guint8 * code;
+  GumArm64Writer cw;
+  gint r;
```
Let's avoid single-letter variable names. Maybe name it `result` or something descriptive.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+  gum_clear_cache (cw.base, gum_arm64_writer_offset (&cw));
+  gum_arm64_writer_clear (&cw);
+
+  GumMemoryRange memory_range;
```
Declaration after statement.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+
+  GumMemoryRange memory_range;
+  memory_range.base_address = GUM_ADDRESS (func_a);
+  memory_range.size = (4 * 20);
```
Redundant parentheses (and they don't add clarity).

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,6 +352,251 @@ store_x0 (GumCpuContext * cpu_context,
+  const gchar * start_lbl = "start";
+
+  fixture->sink->mask = GUM_EXEC;
+  g_assert_cmpuint (fixture->sink->events->len, ==, 0);
```
Let's move these two lines down to right before `func ()`, to make the logic at the end clearer.

oleavr commented on gum/backend-arm64/gumstalker-arm64.c:
```diff
@@ -2570,6 +2594,15 @@ gum_exec_block_write_call_invoke_code (GumExecBlock * block,
+
+    gum_arm64_writer_put_mov_reg_reg (cw, ARM64_REG_X15, ARM64_REG_X0);
+    gum_arm64_writer_put_cbz_reg_label (cw, ARM64_REG_X15, keep_this_blr);
+
```
Extra blank line.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,14 +352,265 @@ store_x0 (GumCpuContext * cpu_context,
+
+  memory_range.base_address = (GumAddress) func_a_address;
+  memory_range.size = 4 * 2;
+
```
Suggest dropping this blank line and the one above, as these are part of the same logical group.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,14 +352,265 @@ store_x0 (GumCpuContext * cpu_context,
+  memory_range.base_address = GUM_ADDRESS (func_a);
+  memory_range.size = 4 * 2;
+
+  gum_stalker_exclude (fixture->stalker, &memory_range);
```
Here as well – though just this blank line.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,14 +352,265 @@ store_x0 (GumCpuContext * cpu_context,
+
+  result = func (2);
+
+  g_assert_cmpint (result, ==, 12);
```
This would be more readable inlined, i.e.:

```C
  g_assert_cmpint (func (2), ==, 12);
```

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,14 +352,265 @@ store_x0 (GumCpuContext * cpu_context,
+  memory_range.base_address = GUM_ADDRESS (func_a);
+  memory_range.size = 4 * 20;
+
+  gum_stalker_exclude (fixture->stalker, &memory_range);
```
Here as well – though just this blank line.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,14 +352,265 @@ store_x0 (GumCpuContext * cpu_context,
+
+  result = func (2);
+
+  g_assert_cmpint (result, ==, 12);
```
This also would be more readable inlined.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,14 +352,265 @@ store_x0 (GumCpuContext * cpu_context,
+
+  result = func (2);
+
+  g_assert_cmpint (result, ==, 12);
```
This also would be more readable inlined.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,14 +352,265 @@ store_x0 (GumCpuContext * cpu_context,
+  memory_range.base_address = GUM_ADDRESS (func_a);
+  memory_range.size = 4 * 20;
+
+  gum_stalker_exclude (fixture->stalker, &memory_range);
```
Here as well – though just this blank line.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -391,9 +648,9 @@ STALKER_TESTCASE (unconditional_branch)
+  result = func (2);
 
-  g_assert_cmpint (r, ==, 13);
+  g_assert_cmpint (result, ==, 13);
```
This also would be more readable inlined.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -449,9 +706,9 @@ STALKER_TESTCASE (unconditional_branch_reg)
+  result = func (2);
 
-  g_assert_cmpint (r, ==, 13);
+  g_assert_cmpint (result, ==, 13);
```
This also would be more readable inlined.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,14 +352,250 @@ store_x0 (GumCpuContext * cpu_context,
+  gum_arm64_writer_put_label (&cw, start_lbl);
+
+  gum_arm64_writer_put_push_reg_reg (&cw, ARM64_REG_X19, ARM64_REG_LR);
+  gum_arm64_writer_put_ldr_reg_address (&cw, ARM64_REG_X1, GUM_ADDRESS (func_a));
```
> 80 cols.

oleavr commented on tests/core/arch-arm64/stalker-arm64.c:
```diff
@@ -346,14 +352,250 @@ store_x0 (GumCpuContext * cpu_context,
+
+  gum_arm64_writer_put_label (&cw, start_lbl);
+  gum_arm64_writer_put_push_reg_reg (&cw, ARM64_REG_X19, ARM64_REG_LR);
+  gum_arm64_writer_put_ldr_reg_address (&cw, ARM64_REG_X1, GUM_ADDRESS (func_a));
```
> 80 cols.

oleavr commented:
Thanks! 🤘
